test-harness.md

```diff
@@ -0,0 +1,94 @@
+MLS Interop Test Harness Specification
+======================================
+
+To automate interoperability testing across multiple MLS implementations, we
+need to be able to have a test runner control multiple clients, on different
+platforms and written in different languages. The interface to the MLS clients
+needs to cover all of the protocol operations for which we want to test interop,
+and expose enough information to evaluate whether interop succeeded.
+
+At a logical level, the necessary operations are all basically RPCs from the
+test runner to the clients.  First we describe the RPCs, then we discuss some
+ways these RPCs could be implemented.
+
+## Test Harness RPCs
+
+Ways to become a member of a group ("encrypt flag" indicates whether handshake
+messages in the group are encrypted):
+
+* Create a new group
+  * Inputs: Group ID, ciphersuite, credential, signature private key, encrypt
+    flag 
+  * Outputs: Pointer to new state
+* Create a new KeyPackage to begin a join
+  * Inputs: ciphersuite, credential, signature private key
+  * Outputs: KeyPackage, callback to which Welcome should be sent
+* Complete joining a group (Welcome callback)
+  * Inputs: Welcome, encrypt flag
+  * Outputs: Pointer to new state
+* External join
+  * Inputs: PublicGroupState, encrypt flag
+  * Outputs: MLSPlaintext(Commit), Welcome, pointer to new state
+
+Operations using a group state:
+
+* Get state auth
+  * Inputs: (none)
+  * Outputs: authentication_secret
+* Export a secret
+  * Inputs: label, context
+  * Outputs: exported_value
+
+* Protect an application message
+  * Inputs: Application data
+  * Outputs: MLSCiphertext(ApplicationData)
+* Unprotect an application message
+  * Inputs: MLSCiphertext(ApplicationData)
+  * Outputs: Application data
+
+* Generate an add proposal
+  * Inputs: KeyPackage
+  * Outputs: MLSPlaintext(Proposal(Add))
+* Generate an update proposal
+  * Inputs: (none)
+  * Outputs: MLSPlaintext(Proposal(Update))
+* Generate an remove proposal
+  * Inputs: index
+  * Outputs: MLSPlaintext(Proposal(Remove))
+* Store an external PSK
+  * Inputs: PSK ID, PSK value
+  * Outputs: (none)
+* Generate a PSK proposal
+  * Inputs: PSK type, PSK ID (depending on type)
+  * Outputs: MLSPlaintext(Proposal(PSK))
+* Generate a ReInit proposal
+  * Inputs: Group ID, ciphersuite
+  * Outputs: MLSPlaintext(Proposal(ReInit))
+* Generate an AppAck proposal for all messages processed in this epoch
+  * Inputs: (none)
+  * Outputs: MLSPlaintext(Proposal(AppAck))
+* Generate a Commit message covering a set of proposals
+  * Inputs: One or more MLSPlaintext(Proposal(\*))
+  * Outputs: Zero or more MLSPlaintext(Proposal(\*)), MLSPlaintext(Commit)
+  * Note: The committer chooses whether to inline any proposals, and returns the
+    proposals not inlined.  If they commiter can't commit one or more of the
+    proposals, it must return an error.
```

Comment:
I think we should be more flexible here. The input proposals don't necessarily have to be from the the committer, they could also be from another sender. In addition, I think the inline proposals should be a separate input.

That way it would cover real-world scenarios where you could have a complete mix of proposals that are covered by a Commit: (own stand-alone proposals, stand-alone proposals from others and inline proposals)

Reply:
Good idea.  I'll update to allow the caller to specify inline / reference proposals separately.